### ISSUE TITLE
SC 2.4.11 Unclear use of thickness, other questions

### ISSUE THREAD

**Issue by yatil** [w3c/wcag#1290]
In 2.4.11 it is imprecise in bullet point 1 what thickness is, and in bullet point 3 what thickness here refers to.

> **Success Criterion 2.4.11 Focus Appearance (Minimum)**
>
> For the keyboard focus indicator of each User Interface Component, all of the following are true:
>
> * **Minimum area:** The focus indication area is greater than or equal to a 1 CSS pixel border of the focused control, or has a thickness of at least 8 CSS pixels along the shortest side of the element.
> * **Change of contrast:** The color change for the focus indication area has a contrast ratio of at least 3:1 with the colors of the unfocused state.
> * **Adjacent contrast:** The focus indication area has a contrast ratio of at least 3:1 against all adjacent colors for the minimum area or greater, or has a thickness of at least 2 CSS pixels.
> * **Unobscured:** The item with focus is not entirely hidden by author-created content.

In bullet point 1 “The focus indication area is greater than or equal to a 1 CSS pixel border of the focused control, or has a thickness of at least 8 CSS pixels along the shortest side of the element.”

I think what is meant is 8px by length of the shortest side of the element, but it could also mean a 8px stripe somewhere along the shortest side of the element. I think this would benefit from simpler, more precise language.

> * **Minimum area:** The focus indication area is greater than or equal to a 1 CSS pixel border of the focused control, or has an at least 8 CSS pixels wide border at the shortest side of the element.

(I still try to process how that relates if my focus indication is not at the shortest side of the element. Could I use a focus indication area that size and align it somewhere else around the element, for example as a partial underline?)

> * **Adjacent contrast:** The focus indication area has a contrast ratio of at least 3:1 against all adjacent colors for the minimum area or greater, or has a thickness of at least 2 CSS pixels.

Do I read that correctly that there are no contrast requirements to the focus indication area when the thickness is at least 2px wide? It still would need a 3:1 contrast ratio to the color it had beforehand, but this could allow a focus indication area that is nearly(?) invisible as long as it is 2px thick.

I _strongly_ think a minimum contrast needs to be ensured for the focus indication area, even if it has 2px width or more.

This point also does not specify that it needs to be a solid border. Would a dotted border, or a dashed border, also be ok to use as a focus indication area?

**Comment by mraccess77** [comment]
HI @yatil we have discussed the 2px size indicators lack of need for contrast in multiple github issues.  In these situations there is still a required differenence from the non-focused state and the size change itself is likely to provide difference.  This provision attempts to allow for situations where it isn't possible to get 3:1 with both internal and external adjacent colors and thus other factors can be used.  While it could be misused I think it provides some flexibility that is needed to ensure we get adoption.  Otherwise we might limit design choices in a way that causes people to ignore the criteria.

**Comment by carrythebanner** [comment]
Related, I'm also having trouble following the numbers that are cited in the SC 2.4.11 "Understanding" doc. 

The "Minimum area" section first says: 
> The following 3 examples use a 90px wide by 30px tall button, therefore the surface area requirement is 240px. 

The first graphical example then says: 
> Figure 2@@@ Update. The inner outline is slightly smaller than the outer edge of the component, but using a 2px thick indicator means it is well over the minimum requirement of 180px squared.

Where does the "minimum requirement of 180px squared" come from?

**Comment by juleskuehn** [comment]
I think "180px squared" is an error and should be replaced by "240px". Referring to areas by units other than linear px (e.g. px squared) should be avoided as it introduces more confusion, in my opinion.

**Comment by alastc** [comment]
Apologies, the "@@@ Update" indicates that it needs updating, in the rush to get everthing out I haven't had a chance to open a graphics editor yet!

**Comment by alastc** [comment]
> or has an at least 8 CSS pixels wide border at the shortest side of the element

I think grammatically that would be:
"or has a border at least 8 CSS pixels wide on the shortest side of the element"

But I'm a little wary of using CSS keywords like "border", it leads people to particular conclusions when we mean it generically. 

> Could I use a focus indication area that size and align it somewhere else around the element, for example as a partial underline?

If it were the length of the shortest side, and at least 8px wide, then yes. It's an alternative calculation.

> Do I read that correctly that there are no contrast requirements to the focus indication area when the thickness is at least 2px wide? It still would need a 3:1 contrast ratio to the color it had beforehand, but this could allow a focus indication area that is nearly(?) invisible as long as it is 2px thick.

From testing, if you add a 2px border around something then it looks significantly thicker/bigger and is (reasonably) visible. E.g. [Example 3 on this test page](https://alastairc.uk/tests/wcag22-examples/focus-more-visible-3.html).

Obviously it would be better if it had contrast against all adjacent colours, but if you've got a solid dark button on a light background, you have a very narrow range of colours (brightnesses) that could contrast with both. Compared to that intermediate colour, it's actually better to have the thickness as it is a better indicator of visibility.

> Would a dotted border, or a dashed border, also be ok to use as a focus indication area?

You can, but you'd need to increase the size as it does have as many pixels compared to a solid line. E.g.  a solid border would need to be 1px/2px depending on contrast. A dotted border has half the pixels, so would need to be 2px or 4px depending on the contrast.

**Comment by yatil** [comment]
Thanks for your responses. It really helps a lot to give adequate suggestions and recommendations for changes, if needed.

>>Would a dotted border, or a dashed border, also be ok to use as a focus indication area?
>
> You can, but you'd need to increase the size as it does have as many pixels compared to a solid line. E.g. a solid border would need to be 1px/2px depending on contrast. A dotted border has half the pixels, so would need to be 2px or 4px depending on the contrast.

I’m a little concerned about this. For one, neither dotted nor dashed has exactly half the pixels, so it would be hard to judge that without getting a ruler out. See how the following codepen renders differently in Safari/Firefox/Edge/Chrome: https://codepen.io/yatil/full/LYNRMwM

I wonder if the pixel measurement is super useful when talking about effects that change the shape of the focus indication area drastically…

**Comment by alastc** [comment]
Well, it's the web, pixels is what we have?!

I did some testing on dotted & dashed borders a while back, and just running your test page through a few browsers and pixel densities:
https://alastairc.uk/tests/wcag22-examples/dot-dash/Edge-1x.png
https://alastairc.uk/tests/wcag22-examples/dot-dash/Edge-2x.png
https://alastairc.uk/tests/wcag22-examples/dot-dash/Firefox-1x.png
https://alastairc.uk/tests/wcag22-examples/dot-dash/Firefox-2x.png
https://alastairc.uk/tests/wcag22-examples/dot-dash/FirefoxMac-2x.png
https://alastairc.uk/tests/wcag22-examples/dot-dash/Safari-2x.png

Dashed is fairly safe at 2:1, i.e. doubling the thickness to get the same area as solid. 
Dotted is a bit more varied, sometimes getting to half the area of the solid line, but sometimes less.

This maps pretty well to how visible they are as well, at least at smaller thicknesses. Perhaps we could add something to the understanding document about avoiding those at the 1-2px thickness?

**Comment by alastc** [comment]
Hi @yatil,

In #1341 and PR #1403 the measurement bullet has been updated to:
> The area of the focus indicator is at least as large as the perimeter of the focused control, as measured by a 1 <a>CSS pixel</a> border, or has a width of at least 8 CSS pixels along the shortest side of the element

Does that help with separating the alternative size measure?

In #1348 I added a suggestion for the anti-aliasing/dotted-lines aspect.

**Comment by mraccess77** [comment]
@alastc I have a question about the alternative - if the person chooses the 8px width - it must actually appear on the shortest side and take up the entire visible height of the control (assuming shortest side is vertical)?  Is that right?  So it seems like it could be either on the left, center or right. (assuming shortest side is vertical)    It also doesn't have to specifically have to be against the edge but be center, inset, etc.      We also aren't calculating area with the 8px - if you aren't using the 8px approach then you need to calculate the 1px border area and then use that for other custom indicators, right?  That is - we could not calculate area as something like 8px times shortest side and then use that area for anything shape indicator as that is not what is meant by the 8px alternative.

**Comment by alastc** [comment]
> we could not calculate area as something like 8px times shortest side and then use that area for anything shape indicator as that is not what is meant by the 8px alternative.

Correct, it is either the area defined in the first part, or 8px * the shortest edge.

**Comment by yatil** [comment]
> The area of the focus indicator is at least as large as the perimeter of the focused control, as measured by a 1 CSS pixel border, or has a width of at least 8 CSS pixels along the shortest side of the element.

Should we specify the shortest side exception/alternative that as a rectangular area:

> The area of the focus indicator is at least as large as the perimeter of the focused control, as measured by a 1 CSS pixel border, or is displayed as a rectangular area that is at least 8 CSS pixels by the number of CSS pixels of the shortest side of the element large.

(I still feel that I’m missing something here… Sorry for my confusion.)

**Comment by mraccess77** [comment]
@alastc  your comment "Correct, it is either the area defined in the first part, or 8px * the shortest edge." indicates that if we use 8px* shortest edge then we are saying it could be anywhere and any thickness.  So could you confirm that we could calculation the area as 8px * shortest edge and then make the indicator be a horizonal line, circle, or shape like a rectangle in the middle of the shape, etc.

**Comment by alastc** [comment]
Hi Jon,

Sorry, I didn't phrase that very well. 

The 8px is not intended to be an area calculation, it has to be an 8px thick change on one of the shortest edges.

**Comment by mraccess77** [comment]
Thanks @alastc I think we should further clarify this 8px thick indicator - it seems like it could be anywhere on the shortest edge right?  It could be in the center of the control as long as it extends from the edges on the shortest side.  Also does the indicator have to be a solid color or could it use multiple colors as long as the colors all contrast?

**Comment by alastc** [comment]
Hi everyone,

Based on an update which was surveyed/discussed in the last two weeks, we've updated the size metric to read:

<hr>

**Minimum area**: The area of the focus indicator is either:

- at least as large as the area of a 1 CSS pixel thick perimeter of the unfocused control;
- at least as large as a 4 CSS pixels border along the shortest side, and no thinner than 2 CSS pixels.

<hr>

Survey: https://www.w3.org/2002/09/wbs/35422/wcag22-focus-appearance-updates2/results#xq1
Minutes: https://www.w3.org/2021/01/19-ag-minutes.html#t02 
Understanding document: https://w3c.github.io/wcag/understanding/focus-appearance-minimum.html (Should update soon)

Based on that we'll close this issue, but please raise a new one if there are problems with the new text.
